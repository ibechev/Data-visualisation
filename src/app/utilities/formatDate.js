export const formatDate = date => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ];
  const tempDate = date.split("-");
  return `${months[parseInt(tempDate[1]) - 1]} ${tempDate[0].substring(2)}`;
};
