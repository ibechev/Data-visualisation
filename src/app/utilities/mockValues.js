export const mockAlerts = [
  {
    name: "Adam Campton",
    event: "end of year review",
    eventDate: "2019-01-18",
    importance: "very"
  },
  {
    name: "Emma O'Sullivan",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Brandon Walsh",
    event: "end of year review",
    eventDate: "2019-02-06",
    importance: "middle"
  },
  {
    name: "David O'Brien",
    event: "end of year review",
    eventDate: "2019-02-23",
    importance: "very"
  },
  {
    name: "Sophie O'Connor",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Sean Connor",
    event: "end of year review",
    eventDate: "2019-01-05",
    importance: "very"
  },
  {
    name: "Gery Brown",
    event: "end of year review",
    eventDate: "2019-01-12",
    importance: "middle"
  },
  {
    name: "Stephen Greenday",
    event: "end of year review",
    eventDate: "2019-02-11",
    importance: "very"
  },
  {
    name: "Lisa Aniston",
    event: "end of year review",
    eventDate: "2018-12-27",
    importance: "low"
  },
  {
    name: "Gordon Niels",
    event: "end of year review",
    eventDate: "2018-12-28",
    importance: "middle"
  }
];

export const mockCustomerInfo = {
  max: 270,
  data: [
    { month: "Dec-2017", value: 99 },
    { month: "Jan-2018", value: 80 },
    { month: "Feb-2018", value: 71 },
    { month: "Mar-2018", value: 96 },
    { month: "Apr-2018", value: 80 },
    { month: "May-2018", value: 115 },
    { month: "Jun-2018", value: 146 },
    { month: "Jul-2018", value: 121 },
    { month: "Aug-2018", value: 160 },
    { month: "Sep-2018", value: 168 },
    { month: "Oct-2018", value: 185 },
    { month: "Nov-2018", value: 203 }
  ]
};
