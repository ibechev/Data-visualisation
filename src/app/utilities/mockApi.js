import { mockAlerts, mockCustomerInfo } from "../utilities/mockValues";

const randomTimeRange = (min, max) => {
  return (Math.floor(Math.random() * (max - min + 1)) + min) * 1000;
};
const simulateLatency = ms => new Promise(resolve => setTimeout(resolve, ms));

export const mockFetchAlerts = async () => {
  await simulateLatency(randomTimeRange(0, 3));
  return mockAlerts.map(alert => ({ ...alert }));
};

export const mockFetchClientsInfo = async () => {
  const { max, data } = mockCustomerInfo;
  await simulateLatency(randomTimeRange(0, 3));
  return {
    max,
    data: data.map(each => ({ ...each }))
  };
};
