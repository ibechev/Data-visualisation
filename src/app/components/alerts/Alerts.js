import React, { Component, Fragment } from "react";
import { mockFetchAlerts } from "../../utilities/mockApi";

import Alert from "./alert/Alert";
import AlertsHeader from "./alertsHeader/AlertsHeader";
import AlertsFooter from "./alertsFooter/AlertsFooter";
import LoadingSpinner from "../loadingSpinner/LoadingSpinner";

class Alerts extends Component {
  constructor(props) {
    super(props);

    this.fetchData = this.fetchData.bind(this);
    this.handleButtonClick = this.handleButtonClick.bind(this);

    this.state = {
      alerts: [],
      loading: true,
      error: false
    };
  }

  componentDidMount() {
    this.fetchData();
  }

  handleButtonClick() {
    const { error, loading } = this.state;
    if (!error && !loading)
      // eslint-disable-next-line
      console.log("Button was clicked");
  }

  async fetchData() {
    await mockFetchAlerts()
      .then(res => {
        this.setState(prevState => ({
          ...prevState,
          alerts: [...res],
          loading: false
        }));
      })
      .catch(err => {
        this.setState(prevState => ({
          ...prevState,
          error: err
        }));
        if (DEVELOPMENT) throw new Error(err);
      });
  }

  render() {
    const { alerts, loading, error } = this.state;

    return (
      <section className="alerts">
        <AlertsHeader
          loading={loading}
          error={error}
          latestAlerts={alerts.length}
          handleSave={this.handleButtonClick}
          handleImportance={this.handleButtonClick}
          handleChart={this.handleButtonClick}
          handleFilter={this.handleButtonClick}
          handleUndo={this.handleButtonClick}
          handleSettings={this.handleButtonClick}
        />

        {error ? (
          <p className="error">Oops, there was a problem</p>
        ) : loading ? (
          <LoadingSpinner title="alerts" />
        ) : (
          <Fragment>
            <ul className="alerts-list">
              {alerts.map((alert, i) => (
                <Alert key={i} {...alert} />
              ))}
            </ul>

            <AlertsFooter
              handleSeeMore={this.handleButtonClick}
              handleViewAll={this.handleButtonClick}
            />
          </Fragment>
        )}
      </section>
    );
  }
}

export default Alerts;
