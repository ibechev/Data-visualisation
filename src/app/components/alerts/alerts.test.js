import React from "react";
import { mount } from "enzyme";
import Alerts from "./Alerts";
import renderer from "react-test-renderer";
import { mockAlerts } from "../../utilities/mockValues";

describe("Expect <Alerts />", () => {
  const defaultState = {
    alerts: [...mockAlerts],
    loading: true,
    error: false
  };

  const setup = (testState = {}) => {
    const state = {
      ...defaultState,
      ...testState
    };
    const wrapper = mount(<Alerts />);
    wrapper.setState({ ...state });
    return wrapper;
  };

  it("to render correctly", () => {
    let wrapper = renderer.create(<Alerts />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to call method 'fetchData()' after the component mounts", () => {
    const spyFetchData = jest.spyOn(Alerts.prototype, "fetchData");
    setup();
    expect(spyFetchData).toHaveBeenCalledTimes(1);
  });

  it("to have 'state.error' set to 'false' and 'state.loading' set to 'true' by default", () => {
    const wrapper = setup();
    expect(wrapper.state("error")).toEqual(false);
    expect(wrapper.state("loading")).toEqual(true);
  });

  describe("when 'state.error' is set to true", () => {
    const testState = {
      error: true
    };

    const wrapper = setup(testState);

    it("to display <p class='error' />", () => {
      expect(wrapper.find("p.error")).toHaveLength(1);
    });

    it("to not display <LoadingSpinner />", () => {
      expect(wrapper.find("LoadingSpinner")).toHaveLength(0);
    });

    it("to not display <ul class='alerts-list' /> or <AlertsFooter />", () => {
      expect(wrapper.find("ul.alert-list")).toHaveLength(0);
      expect(wrapper.find("AlertsFooter")).toHaveLength(0);
    });
  });

  describe("when 'state.error' is set to 'false' and 'state.loading' is set to 'true", () => {
    const wrapper = setup();

    it("to not display <p class='error' />", () => {
      expect(wrapper.find("p.error")).toHaveLength(0);
    });

    it("to not display <ul class='alerts-list' /> or <AlertsFooter />", () => {
      expect(wrapper.find("ul.alert-list")).toHaveLength(0);
      expect(wrapper.find("AlertsFooter")).toHaveLength(0);
    });

    it("to display LoadingSpinner />", () => {
      expect(wrapper.find("LoadingSpinner")).toHaveLength(1);
    });
  });

  describe("when 'state.error' is 'false' and 'state.loading' is 'false'", () => {
    const testState = {
      loading: false
    };

    const wrapper = setup(testState);

    it("to not display <p class='error' />", () => {
      expect(wrapper.find("p.error")).toHaveLength(0);
    });

    it("to not display <LoadingSpinner />", () => {
      expect(wrapper.find("LoadingSpinner")).toHaveLength(0);
    });

    it("to display <AlertsFooter /> and <ul class='alerts-list' /> with the correct amount or <Alert /> components", () => {
      expect(wrapper.find("ul.alerts-list")).toHaveLength(1);
      expect(wrapper.find("AlertsFooter")).toHaveLength(1);
      expect(wrapper.find("Alert")).toHaveLength(defaultState.alerts.length);
    });
  });
});
