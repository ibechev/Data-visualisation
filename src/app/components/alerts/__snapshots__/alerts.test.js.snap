// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Alerts /> to render correctly 1`] = `
<section
  className="alerts"
>
  <header
    className="alerts-header"
  >
    <div
      className="alerts-info"
    >
      <h4
        className="alerts-title"
      >
        Alerts
      </h4>
    </div>
    <div
      className="alerts-controls"
    >
      <div
        className="main-alerts-controls"
      >
        <ul
          className="controls-list"
        >
          <li
            className="control-li"
          >
            <button
              className="button no-body"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-chart-pie"
              />
            </button>
          </li>
          <li
            className="control-li"
          >
            <button
              className="button no-body"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-filter"
              />
            </button>
          </li>
          <li
            className="control-li"
          >
            <button
              className="button no-body"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-undo-alt"
              />
            </button>
          </li>
          <li
            className="control-li"
          >
            <button
              className="button no-body"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-cog"
              />
            </button>
          </li>
        </ul>
        <button
          className="button primary"
          onClick={[Function]}
          type="button"
        >
          Save 
          <i
            className="fas fa-caret-down"
          />
        </button>
      </div>
    </div>
  </header>
  <div
    className="loading-spinner"
  >
    <i
      className="fas fa-spinner fa-spin"
    />
    <p>
      Loading 
      alerts
       ...
    </p>
  </div>
</section>
`;
